### PR TITLE
Hyper v 1709

### PR DESCRIPTION
**Pull request by BoogleCloud**
# Hyper-V Support for Metasploitable 3 Windows 2008 Server
This has been confirmed working on both Windows Server 1709 and Windows 10 Pro 1709. I commented out the Linux VM sections as they appeared incomplete. I would be happy to test and get those working too.

### Hyper-V Switch
If you don't specify an existing External hyper-v switch via the "switch_name" option, Packer will create a new one called "packer-hyperv-iso", however this switch will be Internal type and the VM will never get an IP address nor be able to talk to the Internet. To workaround, either specify an existing External switch with the packer option or change the type of the packer-created one to External and select an appropriate NIC. (You can of course also set VLAN IDs if they are in use in your network.)

### Windows Defender Issue
As of 4/2018, Windows Defender says the included vulnerable version of Jenkins.exe is malware and will delete it. This will cause the packer build to fail out about halfway through provisioning. Theoretically excluding the metasploitable3 directory is sufficient to prevent this.
Add-MpPreference -ExclusionPath "C:\Path\To\metasploitable3"

However, even with the directory excluded I noted that real-time scanning would still remove it when packer copies the resources directory to the VM causing the same unrecoverable error. To prevent this, disable Defender during packer build.
Set-MpPreference -DisableRealtimeMonitoring $true

You can re-enable it once everything is built with:
Set-MpPreference -DisableRealtimeMonitoring $false

### CODE REVIEW

**Comment by jmartin-r7** [comment]
Please remove checkin for the Hyper-V tools ISO.  Below is a possible direct link that could be added to `scripts/installs/vm-guest-tools.bat` to obtain and install the tools in a way that does not require redistribution of the binaries inside the repo.

https://content.deltik.org/mirror/os/windows/vmguest.iso

Note: you will need to `rebase` to completely remove the commit the ISO was in from the branch for it to be mergeable.

**Comment by BoogleCloud** [comment]
Done. Please verify I did the rebase correctly.

Packer has an "iso-urls" variable but looks like it just for failover methods to make sure it gets the main OS install file. I added a section to the build_win2008.ps1 script to validate that vmguest.iso is present and present the URL for download if not.

**Comment by jmartin-r7** [comment]
`build_win2008.ps1` is for use in the host OS and should not download this ISO, please expand `scripts/installs/vm-guest-tools.bat` similar to what is done for vmware-tools already.  The URL can be downloaded, hash validated, extracted using 7zip, and then the installer can be executed from inside the vm packer is creating.  This will change the approach to the provisioner logic requirements as well.

The rebase still contains the ISO in the history. Commit b28fffa and 
the new removal commit 0b7dd5e should be completely removed from the branch commit chain.  Please review methods for `git rebase -i <rapid7_upstream_remote>/master`.

https://git-scm.com/book/en/v2/Git-Branching-Rebasing

**Comment by BoogleCloud** [comment]
Was mulling this over a bit more.

How does Packer know which IP address to connect to with virtualbox/vmware? Are those solutions just better at getting it from the guest OS or possibly getting it through a DHCP on a NAT network?

Other Possible solutions:
Create a NAT switch on the host with powershell for packer to assign to. Normal windows doesn't have a native DHCP server for this so there would either need to be a third party dhcp server (introducing unwanted host changes) or using a shell (local) provisioner to set a matching network configuration. In either case I am not sure how to feed the resulting IP to packer for provisioning.

Determine just the cab file or two for Server 2008 from the vmguest.iso folder and add to the existing first boot image (probably only a megabyte or so).

**Comment by jmartin-r7** [comment]
@BoogleCloud, it sounds like the vm does have network access but packer cannot obtain the address.  In that case is it reasonable to still expect `vm-guest-tools.bat` and include it in the scripts executed during the autounattend phase?  Only adding the file to the floppy for the Hyper-V builder would allow you to add tests in `vm-guest-tools.bat` to avoid double install.

This would address the redistribution concern.

Also I will try and do a full code review of the rest of this tomorrow.

**Comment by BoogleCloud** [comment]
Yes, this seems to work.  I added vm-guest-tools.bat to the setup floppy for the hyper-v builder and then call it with the final command in autounattend (it needs to be last as it triggers a reboot). It gets called with a unique value for %PACKER_BUILD_TYPE% so that it won't install again when packer runs the normal provisioner for the vm tools script.

File: Vagrantfile [hunk 1]
```diff
@@ -2,19 +2,19 @@
 # vi: set ft=ruby :
 
 Vagrant.configure("2") do |config|
-  config.vm.define "ub1404" do |ub1404|
-    ub1404.vm.box = "metasploitable3-ub1404"
-    ub1404.vm.hostname = "metasploitable3-ub1404"
-    config.ssh.username = 'vagrant'
-    config.ssh.password = 'vagrant'
+#  config.vm.define "ub1404" do |ub1404|
```
Comment (jmartin-r7) [inline Vagrantfile]: I am guessing you commented this out during testing.   Please revert this change unless there is some reason is should be removed.

File: Vagrantfile [hunk 2]
```diff
@@ -2,19 +2,19 @@
...
+#    config.ssh.password = 'vagrant'
 
-    ub1404.vm.network "private_network", ip: '172.28.128.3'
-
-    ub1404.vm.provider "virtualbox" do |v|
-      v.name = "Metasploitable3-ub1404"
-      v.memory = 2048
-    end
-  end
+#    ub1404.vm.network "private_network", ip: '172.28.128.3'
```
Comment (jmartin-r7) [inline Vagrantfile]: Same as above, restore unless there needs to be some other change.

File: Vagrantfile [hunk 3]
```diff
@@ -36,9 +41,9 @@ Vagrant.configure("2") do |config|
     end
 
     # Insecure share from the Linux machine
-    win2k8.vm.provision :shell, path: "scripts/installs/install_share_autorun.bat"
-    win2k8.vm.provision :shell, inline: "rm C:\\tmp\\vagrant-shell.bat" # Hack for this bug: https://github.com/mitchellh/vagrant/issues/7614
-    win2k8.vm.provision :shell, path: "scripts/installs/setup_linux_share.bat"
-    win2k8.vm.provision :shell, inline: "rm C:\\tmp\\vagrant-shell.bat" # Hack for this bug: https://github.com/mitchellh/vagrant/issues/7614
+    #win2k8.vm.provision :shell, path: "scripts/installs/install_share_autorun.bat"
```
Comment (jmartin-r7) [inline Vagrantfile]: This was a work around for an issue I am not 100% sure is actually fixed in vagrant.  Unless you tested in macOS and found it is resolved, please restore this and if it causes issue on Hyper-V add a mitigation to keep it when not on Hyper-V.

File: packer/answer_files/2008_r2/Autounattend.xml [hunk 4]
```diff
@@ -272,6 +272,12 @@
                     <Order>99</Order>
                     <RequiresUserInput>true</RequiresUserInput>
                 </SynchronousCommand>
+                <SynchronousCommand wcm:action="add">
+                    <CommandLine>cmd.exe /c "set PACKER_BUILDER_TYPE=hyperv-first-boot&&a:\vm-guest-tools.bat</CommandLine>
```
Comment (jmartin-r7) [inline packer/answer_files/2008_r2/Autounattend.xml]: This is an xml file, the diff here is pointing out that `&` is a special character in most XML parsers. Should these be `&amp;`?

File: packer/templates/windows_2008_r2.json [hunk 5]
```diff
@@ -216,6 +258,8 @@
     "autounattend": "{{template_dir}}/../answer_files/2008_r2/Autounattend.xml",
     "scripts_dir": "{{template_dir}}/../../scripts",
     "resources_dir": "{{template_dir}}/../../resources",
-    "box_version": "0.1.0"
+    "switch_name": "langrill.local",
```
Comment (jmartin-r7) [inline packer/templates/windows_2008_r2.json]: This looks suspiciously like it is specific to your installation, is there a guaranteed valid value to default this too? Can this be removed and a valid default will be automatically selected as virtualbox and vmware builders do?

**Comment by BoogleCloud** [comment]
Corrected issues as remarked (yes, most were due to testing).

Will verify in the morning and hopefully we are good.

Thanks for bearing with me. My current job unfortunately does not require git knowledge and my previous job was an SVN shop so many of the git concepts are new, though they all make sense.!

File: packer/answer_files/2008_r2/Autounattend.xml (hunk 4 above)
Comment (BoogleCloud) [reply packer/answer_files/2008_r2/Autounattend.xml]: Nope, raw & is the way to go. I tried it as &amp; and it doesn't produce the expected result.

**Comment by BoogleCloud** [comment]
The packer build now fully works.

As for the Vagrantfile, I had previously commented out the ubuntu references because as of now the powershell script doesn't have any commands to build it so I figured it was a work-in-progress. I'd be happy to add a hyper-v build target for it this weekend unless there was some reason you hadn't done it yet.

**Comment by BoogleCloud** [comment]
Re-forked this project with changes since May and updated to include the Ubuntu system. See new pull request for Hyper-V 1803.
